Update __torch_dispatch__ to return op overload instead of the opoverload packet function

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72673

1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

**BC Breaking change**
update `func == torch.ops.aten.foo` to 

`func.overloadpacket == torch.ops.aten.foo` (if you want to perform the logical equivalent of the old check) 

`func == torch.ops.aten.foo.my_overload` (if you want to add overload specific behavior)

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

anjali411 committed:
Update __torch_dispatch__ to return op overload instead of the opoverload packet function

[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

zou3519 commented on test/test_autograd.py:
```diff
@@ -7912,7 +7912,7 @@ def __new__(cls, data=None):
             @classmethod
             def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
-                if func == torch.ops.aten.alias:
+                if func.overload_packet == torch.ops.aten.alias:
```
shouldn't this just be `func == torch.ops.aten.alias['']` (or whatever the syntax is for the alias overload)?

anjali411 replied:
sure so we could also change it to `if func == torch.ops.aten.alias.default:` (we use 'default' as the attribute to access '' overload name)

ezyang replied:
Since it took a while for us to make this BC-breaking change, it seems to me that we should introduce some goop to smooth over the transition. The simplest thing we can do is overload the meaning of == on overloads so that they test equal to their overload packet (perhaps throwing a deprecation warning saying that this behavior will get removed in a later version).

We also need to publish guidance on when you should match against the overload_packet (this is still such a weird name to me haha) as opposed to the overload directly. Hot take: matching against the overload packet is probably the wrong thing to do in most cases.

anjali411 replied:
discussed offline that it's ok to just pull the plug

zou3519 commented on torch/_ops.py:
```diff
@@ -46,6 +46,12 @@ def __call__(self, *args, **kwargs):
         return getattr(self._op, key)
 
+    def __eq__(self, other):
+        if self._op == other._op:
```
nit: is self._op "the pointer to a specific operator overload"? If so, it's not clear from the variable name -- self._op_overload may be clearer or maybe a comment next to `self._op = op # pointer to specific operator overload` would help.

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1697,6 +1698,12 @@ void concrete_dispatch_fn(
   // overload resolution but is more complicated (need to expose separate
   // functions per overload)
   py::handle torch_api_function = py::module::import("torch").attr("ops").attr(ns).attr(func_name);
+  py::handle torch_api_function_overload;
```
You should update the comment above that begins with "For now, overloads get coalesced."

zou3519 reviewed (commented):
I like this. Two main things that we should resolve:
1. Are the test failures real?
2. cc @Chillee -- AOTAutograd can be modified to work with this, right?

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1828,7 +1835,7 @@ c10::intrusive_ptr<TensorImpl> concrete_detach_fn(const c10::impl::PyInterpreter
     kwargs.ptr(),
     "detach",
-    py::module::import("torch").attr("ops").attr("aten").attr("detach").ptr(),
+    py::module::import("torch").attr("ops").attr("aten").attr("detach").attr("default").ptr(),
```
huh, I'm surprised this works (granted, I didn't read the opoverload packet PR in detail haha). Does this mean that default is now a reserved keyword for overload names? Do we have tests that ensure that no one tries to name an overload default?

When I look at tools/codegen, I don't see any according logic:

```
ezyang-mbp:pytorch-tmp ezyang$ git grep \"default tools/codegen
tools/codegen/dest/register_dispatch_key.py:    // so this "default" kernel doesn't actually handle backends that don't support at::empty
ezyang-mbp:pytorch-tmp ezyang$ git grep \'default tools/codegen
tools/codegen/gen.py:            arg['default'] = cpp_a.default
tools/codegen/gen.py:        arg['default'] = pythonify_default(cpp.default_expr(a.default, a.type))
tools/codegen/gen.py:        'default': str(f.has_composite_kernel or has_autogenerated_composite_kernel(f))
```

anjali411 replied:
@ezyang I added checks in the schema parser to disable default as an overload name in the schema parser to also disable that when registering through the RegisterOperators API https://github.com/pytorch/pytorch/pull/72206/files

ezyang replied:
OK, I recommend also synchronizing the parser in model.py to also have the same checking; otherwise you get the error only at runtime when you load torch. Doesn't have to be this PR.

ezyang commented on torch/_ops.py:
```diff
@@ -46,6 +46,12 @@ def __call__(self, *args, **kwargs):
+        if self._op == other._op:
+            return True
+        else:
+            return False
```
You don't need an if statement here. Just return `self._op == other._op`. But it would be even better if OpOverload objects were singletons; then you could use ordinary object identity for the equality test here. (But see also BC considerations when there is a type mismatch).

ezyang commented on torch/_ops.py:
```diff
@@ -75,6 +81,12 @@ def __init__(self, qualified_op_name, op_name, op):
+        if self._op == other._op:
+            return True
+        else:
+            return False
```
Ditto here.

ezyang commented:
This broadly looks good to me, although I see there are a number of failing tests. Most of my questions have to do with the prior PR.

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 commented on torch/_ops.py:
```diff
@@ -32,28 +32,28 @@ def __init__(self, overloadpacket, op, schema):
     @property
-    def overload_name(self):
-        return self._schema.overload_name
+    def name(self):
```
does someone have a name suggestion for this?

ezyang replied:
I actually would suggest this as `__str__` and your current `__str__` become `__repr__`

anjali411 replied:
but usually repr should print something that can be used to construct the class object right? This is not true for the current `__str__` return

anjali411 replied:
This property is currently called `qualified_op_name` for the `OpOverloadPacket` btw.

anjali411 replied:
ok so I changed it to `__str__` as you suggested and added angular brackets to the string returned by `__repr__` following the official python guideline: https://docs.python.org/3/library/functions.html#repr

anjali411 commented on torch/_ops.py:
```diff
@@ -32,28 +32,28 @@ def __init__(self, overloadpacket, op, schema):
         self._op = op
         self._schema = schema
         self._overloadpacket = overloadpacket
+        self.__name__ = 'default' if schema.overload_name is '' else schema.overload_name
```
@albanD does this look ok?

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"




[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).

albanD commented on test/test_python_dispatch.py:
```diff
@@ -376,7 +376,7 @@ def __new__(cls, elem, *args, **kwargs):
             @classmethod
             def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
-                if func.__name__ == "clone":
+                if func.overloadpacket.__name__ == "clone":
```
What is `__name__` on the overload func now?

anjali411 replied:
`op_name.overload_name`

albanD commented on torch/_ops.py:
```diff
@@ -32,31 +32,30 @@ def __init__(self, overloadpacket, op, schema):
-    def __str__(self):
-        return "OpOverload(op='{}.{}', overload='{}')".format(*self._schema.name.split("::"), self.overload_name)
+    def __repr__(self):
+        return "<OpOverload(op='{}.{}', overload='{}')>".format(*self._schema.name.split("::"), self.__name__)
```
Why the `<` `>` ?

anjali411 replied:
see https://github.com/pytorch/pytorch/pull/72673#discussion_r819129798

albanD replied:
Makes sense

albanD commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1232,10 +1232,61 @@ void initJITBindings(PyObject* module) {
+                  std::string module_name("torch.ops");
+                  module_name.append(ns);
+                  return pybind11::reinterpret_steal<py::object>(
+                      handle_torch_function_no_python_arg_parser(
```
Why do you want to handle torch function here?

If you do want to, why not do it directly in the python code?

anjali411 replied:
I added this handling to make the `OpOverload`s FX traceable

ezyang replied:
This makes sense. It's a bit bothersome that it's a straight copy paste from the non-overload specific code, curious if we can dedupe a little...

ezyang replied:
Yeah, I read over this and I'm pretty sure it can be deduped. Make a helper function pretty please?

anjali411 replied:
done!

albanD reviewed (commented):
Not sure about the torch function handling, looks good otherwise.

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).

anjali411 committed:
Update on "Update __torch_dispatch__ to return op overload instead of the opoverload packet function"


1. __torch_dispatch__ now returns the OpOverload instead of the OpOverloadPacket to Python.
2. FX can trace the overloads (OpOverload objects) like torch.ops.aten.add.Tensor op, for example
```
import torch.fx as fx
def f(x):
   return torch.ops.aten.add.Tensor(x, x)
print(fx.symbolic_trace(f).code)
```

Differential Revision: [D34627164](https://our.internmc.facebook.com/intern/diff/D34627164)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34627164).
